### PR TITLE
Run CI on hardware

### PR DESCRIPTION
**Pull request by drasmuss**
The main goal here is to allow us to do automated testing and coverage reports on the actual Loihi hardware (prior to this we were just doing this for the emulator).  That all happens in 5840741.

Then there are some other miscellaneous changes that helped make that easier, or came up while I was working on it.

- 28cbaf3: simplifies the way we write/call CI scripts.  Previously it was set up with True/False flags, and then we'd check those flags and call the appropriate script.  One problem with this is that it gave the illusion that you could set multiple flags simultaneously (e.g. `NENGO=true` and `STATIC=true`), when in fact the flags were all mutually exclusive.  Another is that whenever we wanted to add a new script, we had to create a new flag for it and then add a new condition checking for that flag in each step of the build script.  So I changed it so that you just set one top-level variable that determines which script you want to run.  As part of that I changed it so that scripts just ignore build steps that they don't define behaviour for (as opposed to erroring).
- b357acc: update the max nxsdk version we report testing against (since we're now testing against 0.7.5 as part of this automated testing).
- 0865ddc: spent a lot of time looking at travis-ci pytest logs this PR, and this flag just makes them a little easier to parse 😄 
- c57d43f: noticed that this test wasn't working as intended.  It is now working as intended, but it fails, so I marked it as xfail.
- 6497f58: reduces time required to run the test suite.  Builds were taking a really long time on hardware, so I went through various tests and either reduced the run time or reduced the number of conditions we were testing.  All the test behaviour is the same (e.g. if they had time varying inputs, I sped up the inputs so that the  overall input signal was the same), and none of the tolerances changed, so this is a relatively minimal change that reduced the test time significantly.  We could go through and actually change the tests themselves to further reduce test time in the future.
- b0ffbea: noticed that we had a broken link, fixed it. I think this has been broken for a while, but because of the way the test script was set up we weren't actually getting a failing build if the link checker failed so we just didn't notice.

I've tested that the hardware CI works in a couple different ways:
- works with simultaneous builds
- works when rerunning the same build twice
- failing tests result in a failing build
- coverage reports look right (see https://codecov.io/gh/nengo/nengo-loihi/branch/hardware-ci)

If anyone can think of something else I should check, let me know.

### CODE REVIEW

File: .ci/hardware.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+elif [[ "$COMMAND" == "script" ]]; then
+    ssh loihihost << EOF
+        sh /etc/profile
+        sh ~/.bashrc
+        cd /tmp/nengo-loihi-$TRAVIS_JOB_NUMBER
+        conda create -y -n travis-ci-$TRAVIS_JOB_NUMBER python=3.5.2 scipy
+        source activate travis-ci-$TRAVIS_JOB_NUMBER
+        pip install -e .[tests]
+        pip install ~/travis-ci/nxsdk-0.7.5.tar.gz
+        SLURM=1 coverage run -m pytest --target loihi --no-hang -v --duration 50 --color=yes && \
```
Comment (hunse) [inline .ci/hardware.sh]: Is `duration` the same as `durations`? It's listed as the latter in `pytest --help`.

File: .ci/conda.sh [hunk 2]
```diff
@@ -1,16 +1,10 @@
...
 NAME=$0
 COMMAND=$1
 MINICONDA="http://repo.continuum.io/miniconda/Miniconda-latest-Linux-x86_64.sh"
 
-function usage {
-    echo "usage: $NAME <command>"
-    echo
-    echo "  install  Install miniconda and a simple test environment"
-    exit 1
-}
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/conda.sh]: Yeah the main point of this was to give an error message if you called it with the wrong command.  Since in the new framework you can call any script with any command (and they give a warning message if it is a command they don't recognize), that didn't seem necessary anymore.

File: .ci/hardware.sh [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        coverage report -m && \
+        coverage xml
+EOF
+elif [[ "$COMMAND" == "after_success" ]]; then
+    scp loihihost:/tmp/nengo-loihi-$TRAVIS_JOB_NUMBER/coverage.xml coverage.xml
+    eval "bash <(curl -s https://codecov.io/bash)"
+elif [[ "$COMMAND" == "after_script" ]]; then
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/hardware.sh]: This is all set up to match the Travis-CI names for build steps.  I think anyone working with these scripts is going to be familiar with Travis-CI (or they will need to become familiar before understanding these scripts, as a necessary step), so deviating from the Travis-CI conventions is more likely to cause confusion than not.

File: nengo_loihi/tests/test_learning.py [hunk 4]
```diff
@@ -47,13 +48,15 @@ def pes_network(
...
-def test_pes_comm_channel(allclose, plt, seed, Simulator, n_per_dim, dims):
+def test_pes_comm_channel(allclose, plt, seed, Simulator):
+    n_per_dim = 120
+    dims = 3
     tau = 0.01
-    model, probes = pes_network(n_per_dim, dims, seed, learn_synapse=tau)
+    simtime = 0.5
+    model, probes = pes_network(
+        n_per_dim, dims, seed, learn_synapse=tau,
+        learning_rule_type=nengo.PES(learning_rate=1e-2), period=simtime)
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply nengo_loihi/tests/test_learning.py]: You know the learning code better than me, I just wasn't clear what the 1D case was testing that the 3D case wasn't (i.e., what kind of bugs might be caught by the 1D case that wouldn't be caught by the 3D case).  Similarly with the 200 neurons vs 120 neurons, I wasn't sure what different conditions were being tested by those different neuron numbers.

File: nengo_loihi/tests/test_simulator.py [hunk 5]
```diff
@@ -48,13 +48,15 @@ def test_probedict_fallbacks(precompute, Simulator):
     assert conn_bc  # in sim.data
 
 
+@pytest.mark.xfail
```
[the remark this answers is not in the document]
Comment (hunse) [reply nengo_loihi/tests/test_simulator.py]: So lack of #151 is actually not much of a problem, but it still is not trivial to fix this. Particularly with the shorter test time, it means the synapses have much more of an effect and we need to account for those. I think we should have a separate issue/PR to fix this (i.e. leave it as `xfail` for now).

File: setup.py [hunk 6]
```diff
@@ -60,6 +55,12 @@ def read(*filenames, **kwargs):
             "numpydoc>=0.6",
             "sphinx>=1.8",
         ],
+        "tests": [
+            "coverage>=4.3",
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply setup.py]: Yeah, I find it useful to include all the packages we use when running the test suite though, not just the ones directly needed to run the tests.  Otherwise you end up defining some of your dependencies here and some in the script files in Travis-CI and it's just really confusing keeping track of that and keeping it in sync.  Since no one other than developers are likely to be installing the test requirements, it seems alright to me to add some extra dependencies there if it makes development easier.

File: .ci/hardware.sh [hunk 7]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+set -e -v
+
+NAME=$0
+COMMAND=$1
+
+if [[ "$COMMAND" == "install" ]]; then
+    pip install coverage  # needed for the uploading in after_success
+    openssl aes-256-cbc -K $encrypted_e0365add3e93_key -iv $encrypted_e0365add3e93_iv -in .ci/travis_rsa.enc -out ~/.ssh/id_rsa -d
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/hardware.sh]: This is Travis-CI's standard command for decrypting a stored key (it just tells you "copy and paste this command into your script"), it isn't something I developed for this script.  So I don't think it's super critical that we understand how exactly it works, we can just trust Travis-CI that this is the command we should use to decrypt the private key.

I'll add some comments to that effect though.

> How does the public part of this key then get supplied to intelhost?

That is just done manually (by me), the same way you would set up the public keys for your own connections.

File: .ci/hardware.sh [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        sh /etc/profile
+        sh ~/.bashrc
+        cd /tmp/nengo-loihi-$TRAVIS_JOB_NUMBER
+        conda create -y -n travis-ci-$TRAVIS_JOB_NUMBER python=3.5.2 scipy
+        source activate travis-ci-$TRAVIS_JOB_NUMBER
+        pip install -e .[tests]
+        pip install ~/travis-ci/nxsdk-0.7.5.tar.gz
+        SLURM=1 coverage run -m pytest --target loihi --no-hang -v --duration 50 --color=yes && \
+        coverage report -m && \
+        coverage xml
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/hardware.sh]: I believe keeping continuation lines at the same indent level is the bash convention (although I am not an expert).

File: .ci/hardware.sh [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    chmod 600 ~/.ssh/id_rsa
+    echo -e "$INTELHOST_INFO" >> ~/.ssh/config
+    ssh -o StrictHostKeyChecking=no loihihost "echo 'Connected to loihihost'"
+    scp -r . loihihost:/tmp/nengo-loihi-$TRAVIS_JOB_NUMBER
+elif [[ "$COMMAND" == "script" ]]; then
+    ssh loihihost << EOF
+        sh /etc/profile
+        sh ~/.bashrc
+        cd /tmp/nengo-loihi-$TRAVIS_JOB_NUMBER
+        conda create -y -n travis-ci-$TRAVIS_JOB_NUMBER python=3.5.2 scipy
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/hardware.sh]: Yeah, I thought about splitting this up into the different Travis-CI build steps.  But it's already a bit confusing understanding what is happening on which host, so I thought having all the `intelhost` logic happening in one block would be helpful.  Not super attached to it though.

File: .ci/hardware.sh [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+NAME=$0
+COMMAND=$1
+
+if [[ "$COMMAND" == "install" ]]; then
+    pip install coverage  # needed for the uploading in after_success
+    openssl aes-256-cbc -K $encrypted_e0365add3e93_key -iv $encrypted_e0365add3e93_iv -in .ci/travis_rsa.enc -out ~/.ssh/id_rsa -d
+    chmod 600 ~/.ssh/id_rsa
+    echo -e "$INTELHOST_INFO" >> ~/.ssh/config
+    ssh -o StrictHostKeyChecking=no loihihost "echo 'Connected to loihihost'"
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/hardware.sh]: Yeah, and when you're reading the Travis-CI log it's helpful to have a marker for where you are at in the script (since all the above happens silently).

File: .ci/hardware.sh (hunk 1 above)
Comment (drasmuss) [reply .ci/hardware.sh]: Yeah, apparently either one works (I didn't even notice).  I'll change it to durations just for consistency though.

File: .ci/hardware.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (hunse) [reply .ci/hardware.sh]: Does it have to be our convention?

File: .ci/hardware.sh (hunk 1 above)
Comment (hunse) [reply .ci/hardware.sh]: `pytest` is like that. I always use `pytest --plot` to make plots, but the name we actually gave the parameter is `--plots`.

File: .ci/hardware.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (hunse) [reply .ci/hardware.sh]: > This is Travis-CI's standard command for decrypting a stored key (it just tells you "copy and paste this command into your script")

Definitely having the source in the comment would be helpful, then.

File: .ci/hardware.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/hardware.sh]: It doesn't have to, but I prefer it as well :smile:

File: .ci/hardware.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (hunse) [reply .ci/hardware.sh]: I'm not super picky either way, either. I just figured if you're connecting in `install` anyway (to test the connection), might as well do this there. But if you'd have to call `sh /etc/profile; sh ~/.bashrc` in `install` as well, maybe easier just to do this here.

File: .ci/conda.sh [hunk 11]
```diff
@@ -1,16 +1,10 @@
 #!/usr/bin/env bash
+set -e -v
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/conda.sh]: Done!

File: .ci/conda.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/conda.sh]: Added a comment at the top explaining what the point of this script is

File: .ci/docs.sh [hunk 12]
```diff
@@ -1,27 +1,18 @@
...
 
-function usage {
-    echo "usage: $NAME <command>"
-    echo
-    echo "  install  Install dependencies for building docs"
-    echo "  check    Check documentation for broken links"
-    echo "  run      Build documentation"
-    echo "  upload   Upload documentation to gh-pages branch"
-    exit 1
-}
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/docs.sh]: Done!

File: .ci/emulator.sh [hunk 13]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/emulator.sh]: Added a comment explaining the purpose of each script file

File: nengo_loihi/tests/test_learning.py [hunk 14]
```diff
@@ -65,7 +68,7 @@ def test_pes_comm_channel(allclose, plt, seed, Simulator, n_per_dim, dims):
 
     t = nengo_sim.trange()
     pre_tmask = t > 0.1
-    post_tmask = t > simtime - 1.0
+    post_tmask = t > simtime - 0.1
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply nengo_loihi/tests/test_learning.py]: Made it run for two periods, and test for one period (had to increase the runtime from 0.5 -> 0.6)

File: .ci/hardware.sh [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+set -e -v
+
+NAME=$0
+COMMAND=$1
+
+if [[ "$COMMAND" == "install" ]]; then
+    pip install coverage  # needed for the uploading in after_success
+    openssl aes-256-cbc -K $encrypted_e0365add3e93_key -iv $encrypted_e0365add3e93_iv -in .ci/travis_rsa.enc -out ~/.ssh/id_rsa -d
+    chmod 600 ~/.ssh/id_rsa
+    echo -e "$INTELHOST_INFO" >> ~/.ssh/config
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/hardware.sh]: Added comment

File: .ci/hardware.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (drasmuss) [reply .ci/hardware.sh]: Added link to the Travis-CI documentation in comment

File: nengo_loihi/tests/test_learning.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hunse) [reply nengo_loihi/tests/test_learning.py]: I'm not sure what the 1-D case might catch, but sometimes things are different between 1-D and 3-D (e.g. if the error that we send to the chip somehow ends up as a numpy scalar instead of a vector, that could cause a bug). So I think it's worthwhile to test both.

As for the number of neurons, that's less important. I do like varying things just to catch weird things like the 300 bug (really, we should maybe pick random numbers of neurons for our tests, or something), but since this is neurons per dim anyway, we could have it the same for both 1 and 3 dims and have different numbers of neurons.

[the remark this answers is not in the document]
Comment (drasmuss) [reply nengo_loihi/tests/test_learning.py]: Sounds good, added `dims=1,3` cases back in

File: .ci/hardware.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (hunse) [reply .ci/hardware.sh]: I guess these are really independent commands, so maybe it does make sense to have no indent. If we were using pipes `|` instead of `&&` I might argue for intents more.

File: .ci/conda.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (tbekolay) [reply .ci/conda.sh]: For me part of having this message was also if you try to run it with no command, which is a common thing to forget when using it at the command line. I, for example, use `.ci/static.sh run` to run static checks locally so it's IMO a bit of a regression to have a worse error message if I forget to pass in `run`, but probably not worth blocking the merge...

File: .travis.yml [hunk 16]
```diff
@@ -6,44 +6,23 @@ notifications:
...
-    - env: PYTHON="3.5.2"
-    - env: PYTHON="3.5.2" NENGO="true"
-    - env: PYTHON="3.5.2" STATIC="true"
-    - env: PYTHON="3.5.2" DOCS="true"
+    - env: MODE="test"
+    - env: MODE="static"
+    - env: MODE="docs"
 
 before_install:
+  - chmod +x .ci/*.sh
```
Comment (tbekolay) [inline .travis.yml]: We should mark all the scripts as executable in git rather than as part of the build process I think... can do this quickly in the merge.

File: .ci/hardware.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (tbekolay) [reply .ci/hardware.sh]: I also prefer indenting, but yeah since we're doing `&& \` at the end I'm fine with same level. Though I also prefer the binary operator on the following line ;) Anyway, not worth blocking.
